in_emitter: Fix single record chunks and respect mem_buf_limit pause

drbugfinder-work opened the pull request:
This PR covers two issues:
1) This PR corrects the behavior of a reached mem_buf_limit. The plugin used to accept further records even after being paused which leads to https://github.com/fluent/fluent-bit/issues/8198

2) The current `in_emitter` implementation results in ***only one record per chunk*** being created, which is suboptimal. This PR fixes the collector handling: 

Please refer to line:
https://github.com/fluent/fluent-bit/blob/9652b0dc6fd9e59ef87c3bdc59c89da9abed995b/plugins/in_emitter/emitter.c#L165

To validate this observation, please use:
- fluent-bit.conf https://gist.github.com/drbugfinder-work/e5b28ed2c2247c741695eca7c3add110
- multiline_parsers.conf https://gist.github.com/drbugfinder-work/a9979a730fef75242d5cb4058dc93757
- in_emitter_debug_output.diff https://gist.github.com/drbugfinder-work/f5eb5bf175e9a164cf27b954ba41b188

Check the output:

```
[2024/02/09 15:58:22] [trace] [filter:multiline:multiline.0 at /Users/root/Development/fluent-bit-in_emitter/fluent-bit/plugins/filter_multiline/ml.c:177] emitting from dummy.0 to dummy.0
[2024/02/09 15:58:22] [debug] [input:emitter:emitter_for_multiline.0] did not use ring-buffer
[2024/02/09 15:58:22] [debug] [input:emitter:emitter_for_multiline.0] Number of existing chunks: 0
[2024/02/09 15:58:22] [debug] [input:emitter:emitter_for_multiline.0] no candidate chunk found, create a new one for tag: dummy.0
[2024/02/09 15:58:22] [debug] [input:emitter:emitter_for_multiline.0] new chunk created for tag: dummy.0
[2024/02/09 15:58:22] [trace] [filter:multiline:multiline.0 at /Users/root/Development/fluent-bit-in_emitter/fluent-bit/plugins/filter_multiline/ml.c:740] not processing records from the emitter
[2024/02/09 15:58:22] [debug] [input chunk] update output instances with new chunk size diff=205, records=1, input=emitter_for_multiline.0
[2024/02/09 15:58:22] [trace] [filter:multiline:multiline.0 at /Users/root/Development/fluent-bit-in_emitter/fluent-bit/plugins/filter_multiline/ml.c:177] emitting from dummy.0 to dummy.0
[2024/02/09 15:58:22] [debug] [input:emitter:emitter_for_multiline.0] did not use ring-buffer
[2024/02/09 15:58:22] [debug] [input:emitter:emitter_for_multiline.0] Number of existing chunks: 0
[2024/02/09 15:58:22] [debug] [input:emitter:emitter_for_multiline.0] no candidate chunk found, create a new one for tag: dummy.0
[2024/02/09 15:58:22] [debug] [input:emitter:emitter_for_multiline.0] new chunk created for tag: dummy.0
[2024/02/09 15:58:22] [trace] [filter:multiline:multiline.0 at /Users/root/Development/fluent-bit-in_emitter/fluent-bit/plugins/filter_multiline/ml.c:740] not processing records from the emitter

```
Full logs here: 
- output before PR change: https://gist.github.com/drbugfinder-work/d1e216dd89fee32fe6e8239993b3e390
- output after PR change: https://gist.github.com/drbugfinder-work/0c90824d2a947315d3cef5ce4f1f4360



I'm uncertain if this modification is the correct approach to resolve this issue, as it appears that the ring buffer is unused. This PR is partially reverting some parts to older code version.
However, implementing this does address the behavior of only one record being generated per chunk.

Please also see the valgrind output:
https://gist.github.com/drbugfinder-work/83352c59799659db9741f33a22083eaa



---------------------------------------
### _**Other comments, not directly related to this PR:**_



_Additionally, the entire file appears to require restructuring, as some comments are inaccurate, and the `DEFAULT_EMITTER_RING_BUFFER_FLUSH_FREQUENCY`, which I think should represent a time duration, is instead utilized as a 'buffer size'._

 https://github.com/fluent/fluent-bit/blob/9652b0dc6fd9e59ef87c3bdc59c89da9abed995b/plugins/in_emitter/emitter.c#L245


_The scheduler seems to be empty here, so the ring buffer will never be started (proved by another debug output):_

 https://github.com/fluent/fluent-bit/blob/9652b0dc6fd9e59ef87c3bdc59c89da9abed995b/plugins/in_emitter/emitter.c#L241-L259

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:

- [x] Example configuration file for the change
- [x] Debug log output from testing the change
<!--
Please refer to the Developer Guide for instructions on building Fluent Bit with Valgrind support:
https://github.com/fluent/fluent-bit/blob/master/DEVELOPER_GUIDE.md#valgrind
Invoke Fluent Bit and Valgrind as: $ valgrind --leak-check=full ./bin/fluent-bit <args>
-->
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.

- [N/A] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [N/A] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [N/A] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

**Backporting**
<!--
PRs targeting the default master branch will go into the next major release usually.
If this PR should be backported to the current or earlier releases then please submit a PR for that particular branch.
-->
- [N/A] Backport to latest stable release.

<!--  Other release PR (not required but highly recommended for quick turnaround) -->
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

drbugfinder-work commented:
@edsiper @nokute78 @leonardo-albertovich @pwhelan 
Can you please look at this?

drbugfinder-work commented:
@edsiper @nokute78 @agup006

I further noticed, that the in_emitter continues to accepts records even after reaching the mem_buf_limit and being paused. I added a check if the plugin is paused.

This PR should now fix: 
https://github.com/fluent/fluent-bit/issues/8198
https://github.com/fluent/fluent-bit/issues/4049

lecaros commented:
adding to milestone for visibility

nokute78 commented:
@drbugfinder-work I tested following configuration.
I think it drops some inputs. Could you test it ?

```
[INPUT]
    Name head
    Tag input
    file /proc/uptime
    interval_sec 0
    interval_nsec 500000 # polling /proc/uptime every 500us

[FILTER]
    Name rewrite_tag
    Match input
    emitter_mem_buf_limit 1K # small buffer to test buffer full condition
    rule $head ^.+ new_tag false

[OUTPUT]
    Name file
    Match *
```

Output:
```
(snip)
new_tag: [1709342296.675720632, {"head":"9742.69 7888.00\n"}]
new_tag: [1709342296.676202064, {"head":"9742.69 7888.00\n"}]
new_tag: [1709342296.676705789, {"head":"9742.69 7888.00\n"}]
new_tag: [1709342296.677170099, {"head":"9742.69 7888.00\n"}]
new_tag: [1709342296.978255973, {"head":"9743.00 7888.24\n"}]
new_tag: [1709342296.978649250, {"head":"9743.00 7888.24\n"}]
new_tag: [1709342296.979152223, {"head":"9743.00 7888.24\n"}]
new_tag: [1709342296.979651800, {"head":"9743.00 7888.24\n"}]
new_tag: [1709342296.980152078, {"head":"9743.00 7888.24\n"}]
new_tag: [1709342296.980652175, {"head":"9743.00 7888.25\n"}]
(snip)
```

I think following lines mean fluent-bit drop some inputs when in_emitter pauses.
```
new_tag: [1709342296.677170099, {"head":"9742.69 7888.00\n"}]
new_tag: [1709342296.978255973, {"head":"9743.00 7888.24\n"}]
```

drbugfinder-work commented:
Hi @nokute78,

Thanks for your input. The behavior you observed, with potential input drops when the emitter buffer is full, is expected, because your defined head input plugin non-stop provides new data, so there will be of course a skip. 

To address your concern, I replicated your setup and also tested with a tail plugin (**which actually _can_ pause**) configuration to compare behaviors under similar stress conditions:

First generate test data similar like your configuration:
```
[INPUT]
    Name head
    Tag input
    file /proc/uptime
    interval_sec 0
    interval_nsec 500000 # polling /proc/uptime every 500us

[FILTER]
    Name rewrite_tag
    Match input
    emitter_mem_buf_limit 1K # small buffer to test buffer full condition
    rule $head ^.+ new_tag false

[OUTPUT]
    Name file
    Match *
    File input
```
And let it run for 10 seconds.

Then use this configuration with a tail plugin (which can pause) and let it also run for at least 10 seconds:
```
[INPUT]
    Name tail
    Tag input
    Path input
    Read_from_head true

[FILTER]
    Name rewrite_tag
    Match input
    emitter_mem_buf_limit 1K # small buffer to test buffer full condition
    rule $head ^.+ new_tag false

[OUTPUT]
    Name file
    Match *
    File output
```

Now you can compare the input file with the output file and you should see the exact same number of records.

nokute78 commented:
@drbugfinder-work Thank you for information.
I created huge text file that contains 316K lines and use your configuration.
I can see the exact same number of records.
However in_emitter didn't pause and I think it is different from #8198.


```
$ wc test.log 
 316240  632480 5692320 test.log
```

c.conf:
```
[INPUT]
    Name tail
    Tag input
    Path test.log
    Read_from_head true

[FILTER]
    Name rewrite_tag
    Match input
    emitter_mem_buf_limit 1K # small buffer to test buffer full condition
    rule $log ^.+ new_tag false

[OUTPUT]
    Name file
    Match *
    File output
```

```
$ bin/fluent-bit -c c.conf 
Fluent Bit v3.0.0
* Copyright (C) 2015-2024 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

____________________
< Fluent Bit v2.2.2 >
 -------------------
          \
           \
            \          __---__
                    _-       /--______
               __--( /     \ )XXXXXXXXXXX\v.
             .-XXX(   O   O  )XXXXXXXXXXXXXXX-
            /XXX(       U     )        XXXXXXX\
          /XXXXX(              )--_  XXXXXXXXXXX\
         /XXXXX/ (      O     )   XXXXXX   \XXXXX\
         XXXXX/   /            XXXXXX   \__ \XXXXX
         XXXXXX__/          XXXXXX         \__---->
 ---___  XXX__/          XXXXXX      \__         /
   \-  --__/   ___/\  XXXXXX            /  ___--/=
    \-\    ___/    XXXXXX              '--- XXXXXX
       \-\/XXX\ XXXXXX                      /XXXXX
         \XXXXXXXXX   \                    /XXXXX/
          \XXXXXX      >                 _/XXXXX/
            \XXXXX--__/              __-- XXXX/
             -XXXXXXXX---------------  XXXXXX-
                \XXXXXXXXXXXXXXXXXXXXXXXXXX/
                  ""VXXXXXXXXXXXXXXXXXXV""

[2024/03/03 09:12:00] [ info] [fluent bit] version=3.0.0, commit=feb424367d, pid=101988
[2024/03/03 09:12:00] [ info] [storage] ver=1.5.1, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2024/03/03 09:12:00] [ info] [cmetrics] version=0.6.6
[2024/03/03 09:12:00] [ info] [ctraces ] version=0.4.0
[2024/03/03 09:12:00] [ info] [input:tail:tail.0] initializing
[2024/03/03 09:12:00] [ info] [input:tail:tail.0] storage_strategy='memory' (memory only)
[2024/03/03 09:12:00] [ info] [input:emitter:emitter_for_rewrite_tag.0] initializing
[2024/03/03 09:12:00] [ info] [input:emitter:emitter_for_rewrite_tag.0] storage_strategy='memory' (memory only)
[2024/03/03 09:12:00] [ info] [sp] stream processor started
[2024/03/03 09:12:00] [ info] [output:file:file.0] worker #0 started
[2024/03/03 09:12:01] [ info] [input:tail:tail.0] inotify_fs_add(): inode=4181157 watch_fd=1 name=test.log
^C[2024/03/03 09:12:07] [engine] caught signal (SIGINT)
[2024/03/03 09:12:07] [ warn] [engine] service will shutdown in max 5 seconds
[2024/03/03 09:12:07] [ info] [input] pausing tail.0
[2024/03/03 09:12:07] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/03 09:12:07] [ info] [engine] service has stopped (0 pending tasks)
[2024/03/03 09:12:07] [ info] [input] pausing tail.0
[2024/03/03 09:12:07] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/03 09:12:07] [ info] [output:file:file.0] thread worker #0 stopping...
[2024/03/03 09:12:07] [ info] [output:file:file.0] thread worker #0 stopped
[2024/03/03 09:12:07] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=4181157 watch_fd=1
```

```
$ wc output 
  316240  1264960 19290640 output
```

nokute78 commented:
I also tested following configuration and script.
The script creates 1000 lines log and it output log every 500us.
in_emitter paused and input log lines and output log lines were different.

a.sh:
```sh
#!/bin/bash

for i in `seq 1 1000`
do
    sleep 0.0005
    cat /proc/uptime >> test.log
done
```

b.conf:
```
[INPUT]
    Name tail
    Tag input
    path test.log

[FILTER]
    Name rewrite_tag
    Match input
    emitter_mem_buf_limit 1K
    rule $log ^.+ new_tag false

[OUTPUT]
    Name file
    Match *
```

```
$ bin/fluent-bit -c b.conf 
Fluent Bit v3.0.0
* Copyright (C) 2015-2024 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

____________________
< Fluent Bit v2.2.2 >
 -------------------
          \
           \
            \          __---__
                    _-       /--______
               __--( /     \ )XXXXXXXXXXX\v.
             .-XXX(   O   O  )XXXXXXXXXXXXXXX-
            /XXX(       U     )        XXXXXXX\
          /XXXXX(              )--_  XXXXXXXXXXX\
         /XXXXX/ (      O     )   XXXXXX   \XXXXX\
         XXXXX/   /            XXXXXX   \__ \XXXXX
         XXXXXX__/          XXXXXX         \__---->
 ---___  XXX__/          XXXXXX      \__         /
   \-  --__/   ___/\  XXXXXX            /  ___--/=
    \-\    ___/    XXXXXX              '--- XXXXXX
       \-\/XXX\ XXXXXX                      /XXXXX
         \XXXXXXXXX   \                    /XXXXX/
          \XXXXXX      >                 _/XXXXX/
            \XXXXX--__/              __-- XXXX/
             -XXXXXXXX---------------  XXXXXX-
                \XXXXXXXXXXXXXXXXXXXXXXXXXX/
                  ""VXXXXXXXXXXXXXXXXXXV""

[2024/03/03 09:22:20] [ info] [fluent bit] version=3.0.0, commit=feb424367d, pid=116425
[2024/03/03 09:22:20] [ info] [storage] ver=1.5.1, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2024/03/03 09:22:20] [ info] [cmetrics] version=0.6.6
[2024/03/03 09:22:20] [ info] [ctraces ] version=0.4.0
[2024/03/03 09:22:20] [ info] [input:tail:tail.0] initializing
[2024/03/03 09:22:20] [ info] [input:tail:tail.0] storage_strategy='memory' (memory only)
[2024/03/03 09:22:20] [ info] [input:emitter:emitter_for_rewrite_tag.0] initializing
[2024/03/03 09:22:20] [ info] [input:emitter:emitter_for_rewrite_tag.0] storage_strategy='memory' (memory only)
[2024/03/03 09:22:20] [ info] [sp] stream processor started
[2024/03/03 09:22:20] [ info] [input:tail:tail.0] inotify_fs_add(): inode=4180947 watch_fd=1 name=test.log
[2024/03/03 09:22:20] [ info] [output:file:file.0] worker #0 started
[2024/03/03 09:22:27] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/03 09:22:27] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/03 09:22:28] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/03 09:22:28] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/03 09:22:28] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/03 09:22:29] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/03 09:22:29] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/03 09:22:29] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/03 09:22:30] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/03 09:22:30] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/03 09:22:30] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/03 09:22:31] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/03 09:22:31] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/03 09:22:31] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/03 09:22:32] [ info] [input] emitter.1 resume (mem buf overlimit)
^C[2024/03/03 09:22:34] [engine] caught signal (SIGINT)
[2024/03/03 09:22:34] [ warn] [engine] service will shutdown in max 5 seconds
[2024/03/03 09:22:34] [ info] [input] pausing tail.0
[2024/03/03 09:22:34] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/03 09:22:35] [ info] [engine] service has stopped (0 pending tasks)
[2024/03/03 09:22:35] [ info] [input] pausing tail.0
[2024/03/03 09:22:35] [ info] [output:file:file.0] thread worker #0 stopping...
[2024/03/03 09:22:35] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/03 09:22:35] [ info] [output:file:file.0] thread worker #0 stopped
[2024/03/03 09:22:35] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=4180947 watch_fd=1
```


Log lines are different.
```
$ wc new_tag test.log 
  116   464  7076 new_tag
 1000  2000 18000 test.log
 1116  2464 25076 total
```

braydonk commented:
When I run both of the scenarios that use a fully generated input file read at once, I get equal line counts for each. When I run the scenario with the `head` plugin generating records from proc uptime, I get no pauses. When I run the second scenario, I pretty consistently get one pause, but always seem to get the right number of records. Here's the debug output:
```
Fluent Bit v3.0.0
* Copyright (C) 2015-2024 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

____________________
< Fluent Bit v2.2.2 >
 -------------------
          \
           \
            \          __---__
                    _-       /--______
               __--( /     \ )XXXXXXXXXXX\v.
             .-XXX(   O   O  )XXXXXXXXXXXXXXX-
            /XXX(       U     )        XXXXXXX\
          /XXXXX(              )--_  XXXXXXXXXXX\
         /XXXXX/ (      O     )   XXXXXX   \XXXXX\
         XXXXX/   /            XXXXXX   \__ \XXXXX
         XXXXXX__/          XXXXXX         \__---->
 ---___  XXX__/          XXXXXX      \__         /
   \-  --__/   ___/\  XXXXXX            /  ___--/=
    \-\    ___/    XXXXXX              '--- XXXXXX
       \-\/XXX\ XXXXXX                      /XXXXX
         \XXXXXXXXX   \                    /XXXXX/
          \XXXXXX      >                 _/XXXXX/
            \XXXXX--__/              __-- XXXX/
             -XXXXXXXX---------------  XXXXXX-
                \XXXXXXXXXXXXXXXXXXXXXXXXXX/
                  ""VXXXXXXXXXXXXXXXXXXV""

[2024/03/04 21:57:01] [ info] Configuration:
[2024/03/04 21:57:01] [ info]  flush time     | 1.000000 seconds
[2024/03/04 21:57:01] [ info]  grace          | 5 seconds
[2024/03/04 21:57:01] [ info]  daemon         | 0
[2024/03/04 21:57:01] [ info] ___________
[2024/03/04 21:57:01] [ info]  inputs:
[2024/03/04 21:57:01] [ info]      tail
[2024/03/04 21:57:01] [ info] ___________
[2024/03/04 21:57:01] [ info]  filters:
[2024/03/04 21:57:01] [ info]      rewrite_tag.0
[2024/03/04 21:57:01] [ info] ___________
[2024/03/04 21:57:01] [ info]  outputs:
[2024/03/04 21:57:01] [ info]      file.0
[2024/03/04 21:57:01] [ info] ___________
[2024/03/04 21:57:01] [ info]  collectors:
[2024/03/04 21:57:01] [ info] [fluent bit] version=3.0.0, commit=326ac0c7ee, pid=1619761
[2024/03/04 21:57:01] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2024/03/04 21:57:01] [ info] [storage] ver=1.5.1, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2024/03/04 21:57:01] [ info] [cmetrics] version=0.6.6
[2024/03/04 21:57:01] [ info] [ctraces ] version=0.4.0
[2024/03/04 21:57:01] [ info] [input:tail:tail.0] initializing
[2024/03/04 21:57:01] [ info] [input:tail:tail.0] storage_strategy='memory' (memory only)
[2024/03/04 21:57:01] [debug] [tail:tail.0] created event channels: read=21 write=22
[2024/03/04 21:57:01] [debug] [input:tail:tail.0] flb_tail_fs_inotify_init() initializing inotify tail input
[2024/03/04 21:57:01] [debug] [input:tail:tail.0] inotify watch fd=27
[2024/03/04 21:57:01] [debug] [input:tail:tail.0] scanning path test.log
[2024/03/04 21:57:01] [debug] [input:tail:tail.0] inode=3015869 with offset=0 appended as test.log
[2024/03/04 21:57:01] [debug] [input:tail:tail.0] scan_glob add(): test.log, inode 3015869
[2024/03/04 21:57:01] [debug] [input:tail:tail.0] 1 new files found on path 'test.log'
[2024/03/04 21:57:01] [ info] [input:emitter:emitter_for_rewrite_tag.0] initializing
[2024/03/04 21:57:01] [ info] [output:file:file.0] worker #0 started
[2024/03/04 21:57:01] [ info] [input:emitter:emitter_for_rewrite_tag.0] storage_strategy='memory' (memory only)
[2024/03/04 21:57:01] [debug] [emitter:emitter_for_rewrite_tag.0] created event channels: read=29 write=30
[2024/03/04 21:57:01] [debug] [file:file.0] created event channels: read=31 write=32
[2024/03/04 21:57:01] [debug] [router] match rule tail.0:file.0
[2024/03/04 21:57:01] [debug] [router] match rule emitter.1:file.0
[2024/03/04 21:57:01] [ info] [sp] stream processor started
[2024/03/04 21:57:01] [debug] [input:tail:tail.0] [static files] processed 22.5K
[2024/03/04 21:57:01] [debug] [input chunk] update output instances with new chunk size diff=57000, records=1000, input=emitter_for_rewrite_tag.0
[2024/03/04 21:57:01] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/04 21:57:01] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 21:57:01] [debug] [input:tail:tail.0] inode=3015869 file=test.log promote to TAIL_EVENT
[2024/03/04 21:57:01] [ info] [input:tail:tail.0] inotify_fs_add(): inode=3015869 watch_fd=1 name=test.log
[2024/03/04 21:57:01] [debug] [input:tail:tail.0] [static files] processed 0b, done
[2024/03/04 21:57:02] [debug] [task] created task=0x7f3cf80465d0 id=0 OK
[2024/03/04 21:57:02] [debug] [output:file:file.0] task_id=0 assigned to thread #0
[2024/03/04 21:57:02] [debug] [out flush] cb_destroy coro_id=0
[2024/03/04 21:57:02] [debug] [task] destroy task=0x7f3cf80465d0 (task_id=0)
[2024/03/04 21:57:02] [ info] [input] emitter.1 resume (mem buf overlimit)
^C[2024/03/04 21:57:11] [engine] caught signal (SIGINT)
[2024/03/04 21:57:11] [ warn] [engine] service will shutdown in max 5 seconds
[2024/03/04 21:57:11] [ info] [input] pausing tail.0
[2024/03/04 21:57:11] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 21:57:12] [ info] [engine] service has stopped (0 pending tasks)
[2024/03/04 21:57:12] [ info] [input] pausing tail.0
[2024/03/04 21:57:12] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 21:57:12] [ info] [output:file:file.0] thread worker #0 stopping...
[2024/03/04 21:57:12] [ info] [output:file:file.0] thread worker #0 stopped
[2024/03/04 21:57:12] [debug] [input:tail:tail.0] inode=3015869 removing file name test.log
[2024/03/04 21:57:12] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=3015869 watch_fd=1
```
And the result is:
```
$ wc -l test.log new_tag 
 1000 test.log
 1000 new_tag
 2000 total
```
Note I did add `Read_from_head` which the example config in the latest config didn't have.

drbugfinder-work commented:
Hi @nokute78,
(cc Hi @braydonk),


I tested the same way, but get different results:
1) With `a.sh` being executed first and then Fluent Bit runs with `tail` plugin and `Read_from_head true` option set, the output looks like this with same output numbers:
```
root@fb-inemitter:~/fluent-bit-in_emitter# ./a.sh 
root@fb-inemitter:~/fluent-bit-in_emitter# ls
a.sh  fluent-bit  fluent-bit.conf  test1  test.log
root@fb-inemitter:~/fluent-bit-in_emitter# fluent-bit/build/bin/fluent-bit -c fluent-bit.conf 
...
[2024/03/04 09:58:30] [ info] [input:emitter:emitter_for_rewrite_tag.0] initializing
[2024/03/04 09:58:30] [ info] [input:emitter:emitter_for_rewrite_tag.0] storage_strategy='memory' (memory only)
[2024/03/04 09:58:30] [ info] [sp] stream processor started
[2024/03/04 09:58:30] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/04 09:58:30] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 09:58:30] [ info] [input:tail:tail.0] inotify_fs_add(): inode=304352 watch_fd=1 name=test.log
[2024/03/04 09:58:30] [ info] [output:file:file.0] worker #0 started
[2024/03/04 09:58:31] [ info] [input] emitter.1 resume (mem buf overlimit)
^C[2024/03/04 09:59:26] [engine] caught signal (SIGINT)
[2024/03/04 09:59:26] [ warn] [engine] service will shutdown in max 5 seconds
[2024/03/04 09:59:26] [ info] [input] pausing tail.0
[2024/03/04 09:59:26] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 09:59:27] [ info] [engine] service has stopped (0 pending tasks)
[2024/03/04 09:59:27] [ info] [input] pausing tail.0
[2024/03/04 09:59:27] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 09:59:27] [ info] [output:file:file.0] thread worker #0 stopping...
[2024/03/04 09:59:27] [ info] [output:file:file.0] thread worker #0 stopped
[2024/03/04 09:59:27] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=304352 watch_fd=1
root@fb-inemitter:~/fluent-bit-in_emitter# wc -l new_tag test.log 
 1000 new_tag
 1000 test.log
 2000 insgesamt
```

2) With first Fluent Bit started and then let a.sh run, Fluent Bit stops and resumes the plugin more often and the numbers differ in the end (808 vs 1000):

```
root@fb-inemitter:~/fluent-bit-in_emitter# fluent-bit/build/bin/fluent-bit -c fluent-bit.conf 
...
[2024/03/04 10:39:48] [ info] [sp] stream processor started
[2024/03/04 10:39:48] [ info] [input:tail:tail.0] inotify_fs_add(): inode=304352 watch_fd=1 name=test.log
[2024/03/04 10:39:48] [ info] [output:file:file.0] worker #0 started
[2024/03/04 10:39:57] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/04 10:39:57] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 10:39:58] [ info] [input] emitter.1 resume (mem buf overlimit)
...
[2024/03/04 10:40:36] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/04 10:40:36] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 10:40:37] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/04 10:40:41] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/04 10:40:41] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 10:40:42] [ info] [input] emitter.1 resume (mem buf overlimit)
^C[2024/03/04 11:06:50] [engine] caught signal (SIGINT)
[2024/03/04 11:06:50] [ warn] [engine] service will shutdown in max 5 seconds
[2024/03/04 11:06:50] [ info] [input] pausing tail.0
[2024/03/04 11:06:50] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 11:06:51] [ info] [engine] service has stopped (0 pending tasks)
[2024/03/04 11:06:51] [ info] [input] pausing tail.0
[2024/03/04 11:06:51] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 11:06:51] [ info] [output:file:file.0] thread worker #0 stopping...
[2024/03/04 11:06:51] [ info] [output:file:file.0] thread worker #0 stopped
[2024/03/04 11:06:51] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=304352 watch_fd=1
root@fb-inemitter:~/fluent-bit-in_emitter# wc -l new_tag test.log 
  808 new_tag
 1000 test.log
 1808 insgesamt
```

3) However, if you set the log level to debug/trace you will see expected output:
```
[2024/03/04 11:08:45] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/04 11:08:45] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/04 11:08:45] [debug] [input:tail:tail.0] inode=304352, test.log, events: IN_MODIFY 
[2024/03/04 11:08:45] [debug] [input:emitter:emitter_for_rewrite_tag.0] emitter memory buffer limit reached. 
Not accepting record.
[2024/03/04 11:08:45] [debug] [input:emitter:emitter_for_rewrite_tag.0] emitter memory buffer limit reached. 
Not accepting record.
[2024/03/04 11:08:45] [debug] [input:tail:tail.0] inode=304352, test.log, events: IN_MODIFY 
[2024/03/04 11:08:45] [debug] [input:emitter:emitter_for_rewrite_tag.0] emitter memory buffer limit reached. 
Not accepting record.
[2024/03/04 11:08:45] [debug] [input:tail:tail.0] inode=304352, test.log, events: IN_MODIFY 
...
[2024/03/04 11:08:46] [debug] [input:tail:tail.0] inode=304352, test.log, events: IN_MODIFY 
[2024/03/04 11:08:46] [debug] [input:emitter:emitter_for_rewrite_tag.0] emitter memory buffer limit reached. 
Not accepting record.
[2024/03/04 11:08:46] [trace] [task 0x7fb95a63bbe0] created (id=0)
[2024/03/04 11:08:46] [debug] [task] created task=0x7fb95a63bbe0 id=0 OK
[2024/03/04 11:08:46] [debug] [output:file:file.0] task_id=0 assigned to thread #0
[2024/03/04 11:08:46] [debug] [out flush] cb_destroy coro_id=3
[2024/03/04 11:08:46] [trace] [coro] destroy coroutine=0x7fb959010000 data=0x7fb959010020
[2024/03/04 11:08:46] [trace] [engine] [task event] task_id=0 out_id=0 return=OK
[2024/03/04 11:08:46] [debug] [task] destroy task=0x7fb95a63bbe0 (task_id=0)
[2024/03/04 11:08:46] [ info] [input] emitter.1 resume (mem buf overlimit)

```

This is due to the fact that the tail input plugin is not getting paused - although it should. This situation should never occur. So there's still the missing feedback to the input(s), when the `in_emitter` is paused, which I've talked about in the `Community Meeting #50` (and in issue #8198).

Therefore I added a logic (see second commit) to notify the `in_emitter` about the source input plugin. Potentially, there can be multiple inputs sending to one `in_emitter`, so all the `in_emitter` instance keeps track of them in a list. 
The `in_emitter` will then pause/resume these inputs, too, as soon as the `in_emitter` is being paused/resumed. I also added a unit-test for this in the `multiline` tests - it looks promising. 
However, this mechanism still needs more testing and review. Can you please look at the code @nokute78 @lecaros @braydonk (cc @edsiper) and test it from your end?

Thank you all!

nokute78 commented:
@drbugfinder-work @braydonk  Thank you for updating/testing.
I tested a new patch and it worked on case2 "With first Fluent Bit started and then let a.sh run".

However this approach will not work for some input plugins that don't support pause/resume callbacks. (attach config)
I think we can take care another PR to append these callbacks.

Throughput can be less.
Some records should be passed through if the rule of rewrite_tag is not matched.
However this patch pause input plugins and it means paused plugin can't generate such record.
I don't have any idea for it.

@edsiper Do you have any comments about pausing other input plugins by in_emitter ?
Will race issue happen since engine and in_emitter pause input plugins ?

## testing for in_http that doesn't support pause/resume callback

d.conf:
```
[INPUT]
    Name http
    port 8888

[FILTER]
    Name rewrite_tag
    Match input
    emitter_mem_buf_limit 1K
    rule $log ^.+ new_tag false

[OUTPUT]
    Name file
    Match *
```

b.sh:
```sh
#!/bin/bash

for i in `seq 1 1000`
do
    sleep 0.0005
    UPTIME=`cat /proc/uptime | cut -f1 -d ' '`
    CURL_JSON="{\"log\":\"$UPTIME\"}"
    curl -d $CURL_JSON -XPOST -H "content-type: application/json" http://localhost:8888/input
    echo $CURL_JSON >> test.log
done
```

```
./fluent-bit -c d.conf &
./b.sh
wc test.log new_tag
```

Output:
```
$ wc new_tag test.log 
  445  1335 22695 new_tag
 1000  1000 18000 test.log
 1445  2335 40695 total
```

```
$ bin/fluent-bit -c d.conf 
Fluent Bit v3.0.0
* Copyright (C) 2015-2024 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

(snip)

[2024/03/09 09:29:40] [ info] [fluent bit] version=3.0.0, commit=feb424367d, pid=34012
[2024/03/09 09:29:40] [ info] [storage] ver=1.5.1, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2024/03/09 09:29:40] [ info] [cmetrics] version=0.6.6
[2024/03/09 09:29:40] [ info] [ctraces ] version=0.4.0
[2024/03/09 09:29:40] [ info] [input:http:http.0] initializing
[2024/03/09 09:29:40] [ info] [input:http:http.0] storage_strategy='memory' (memory only)
[2024/03/09 09:29:40] [ info] [input:emitter:emitter_for_rewrite_tag.0] initializing
[2024/03/09 09:29:40] [ info] [input:emitter:emitter_for_rewrite_tag.0] storage_strategy='memory' (memory only)
[2024/03/09 09:29:40] [ info] [sp] stream processor started
[2024/03/09 09:29:40] [ info] [output:file:file.0] worker #0 started
[2024/03/09 09:29:48] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:48] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:48] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:48] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:29:49] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:49] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:49] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:49] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:29:50] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:50] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:50] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:50] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:29:51] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:51] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:51] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:51] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:29:52] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:52] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:52] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:52] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:29:53] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:53] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:53] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:53] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:29:54] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:54] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:54] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:54] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:29:55] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:55] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:55] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:55] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:29:56] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:56] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:56] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:56] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:29:57] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:57] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:57] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:57] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:29:58] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:58] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:58] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:58] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:29:59] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:29:59] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:29:59] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:29:59] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:30:00] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:30:00] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:30:00] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:30:00] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:30:01] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:30:01] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:30:01] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:30:01] [ info] [input] emitter.1 resume (mem buf overlimit)
[2024/03/09 09:30:02] [ warn] [input] emitter.1 paused (mem buf overlimit)
[2024/03/09 09:30:02] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:30:02] [ info] [input] resume emitter_for_rewrite_tag.0
[2024/03/09 09:30:02] [ info] [input] emitter.1 resume (mem buf overlimit)
^C[2024/03/09 09:30:04] [engine] caught signal (SIGINT)
[2024/03/09 09:30:04] [ warn] [engine] service will shutdown in max 5 seconds
[2024/03/09 09:30:04] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:30:04] [ info] [engine] service has stopped (0 pending tasks)
[2024/03/09 09:30:04] [ info] [output:file:file.0] thread worker #0 stopping...
[2024/03/09 09:30:04] [ info] [input] pausing emitter_for_rewrite_tag.0
[2024/03/09 09:30:04] [ info] [output:file:file.0] thread worker #0 stopped
```

drbugfinder-work commented:
Hi @nokute78,

thanks for mentioning this - I didn't think of plugins which cannot be paused.
On the other hand, in case the input plugin cannot be paused, the records (with this PR implemented) will just be dropped - which *should* be the same result as like the inputs (like in_http) would be paused and simply does not accept any new records. In both cases these records would be lost.

> I think we can take care another PR to append these callbacks.
Agree. As we discussed in the issue #8198 the in_http plugin should for example return HTTP 503. 

> However this patch pause input plugins and it means paused plugin can't generate such record.
I think this is a side-effect which cannot be solved in the current pipeline implementation. 

> Will race issue happen since engine and in_emitter pause input plugins ?
I couldn't observe any side-effects, but as I wrote, I'm not sure if a plugin could be removed/exited during runtime (or only during shutdown).

jsubirat commented:
Hey folks, we have tested this PR branch by generating the Docker image out of it and tested it in a K8s cluster that was affected by this issue and have observed very positive results, it seems to fix the issue. Is there an expected timeline to get it merged? Thanks

bdeetz commented:
While rolling fluentbit out in my environment, we ran into this bug. We have built and run fluentbit off of this PR with no sign of failure. It has been moving 10+TB/day of logs for 12 days across 500-1500 collectors. We would very much like to see a fix, if not this one, introduced in order to avoid log loss.

drbugfinder-work commented:
> @drbugfinder-work thanks for triaging and fixing this. In order to merge it, would you please split the commits per component ? ,e.g:
> 
> * filter_multiline: ...
> * filter_rewrite_tag: ...
> * ..
> 
> thanks

@edsiper Done! I've split the commits according to components.
